<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="SWSchaeffer" />

<meta name="date" content="2020-10-03" />

<title>GeneConversionInversion</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GeneConversionInversion</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/SWSchaeffer/GeneConversionInversion">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">GeneConversionInversion</h1>
<h4 class="author">SWSchaeffer</h4>
<h4 class="date">2020-10-03</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-10-05
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>GeneConversionInversion/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201003code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201003)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201003code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201003)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomSWSchaefferGeneConversionInversiontree894c4c0835cb64486c259dac7cbde8cb02097e00targetblank894c4c0a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/SWSchaeffer/GeneConversionInversion/tree/894c4c0835cb64486c259dac7cbde8cb02097e00" target="_blank">894c4c0</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomSWSchaefferGeneConversionInversiontree894c4c0835cb64486c259dac7cbde8cb02097e00targetblank894c4c0a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/SWSchaeffer/GeneConversionInversion/tree/894c4c0835cb64486c259dac7cbde8cb02097e00" target="_blank">894c4c0</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  data/GeneConversion_Check.tsv
    Untracked:  data/GeneConversion_Sites_Reg01.tsv
    Untracked:  data/GeneConversion_Sites_Reg02.tsv
    Untracked:  data/GeneConversion_Sites_Reg03.tsv
    Untracked:  data/GeneConversion_Sites_Reg04.tsv
    Untracked:  data/GeneConversion_Sites_Reg05.tsv
    Untracked:  data/GeneConversion_Sites_Reg06.tsv
    Untracked:  data/GeneConversion_Sites_Reg07.tsv
    Untracked:  data/GeneConversion_Sites_Reg08.tsv
    Untracked:  data/GeneConversion_Sites_Reg09.tsv
    Untracked:  data/GeneConversion_Sites_Reg10.tsv
    Untracked:  data/GeneConversion_Sites_Reg11.tsv
    Untracked:  data/GeneConversion_Sites_Reg12.tsv
    Untracked:  data/GeneConversion_Sites_Reg13.tsv
    Untracked:  data/GeneConversion_Sites_Reg14.tsv
    Untracked:  data/GeneConversion_Stats.tsv
    Untracked:  data/GeneConversion_Tracts.tsv
    Untracked:  data/desktop.ini

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/GeneConversionInversion.Rmd</code>) and HTML (<code>docs/GeneConversionInversion.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/SWSchaeffer/GeneConversionInversion/blob/894c4c0835cb64486c259dac7cbde8cb02097e00/analysis/GeneConversionInversion.Rmd" target="_blank">894c4c0</a>
</td>
<td>
SWSchaeffer
</td>
<td>
2020-10-05
</td>
<td>
wflow_publish(c(“data/<em>.txt", "output/</em>.csv”, “analysis/GeneConversionInversion.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/SWSchaeffer/GeneConversionInversion/74207b78b8fc01d3e92800ba5c292a116d9d6742/docs/GeneConversionInversion.html" target="_blank">74207b7</a>
</td>
<td>
SWSchaeffer
</td>
<td>
2020-10-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/SWSchaeffer/GeneConversionInversion/blob/3428e0929da06d7feb79bc7c01b7bfe178cf0092/analysis/GeneConversionInversion.Rmd" target="_blank">3428e09</a>
</td>
<td>
SWSchaeffer
</td>
<td>
2020-10-03
</td>
<td>
Publish the initial files for GeneConversionIversion
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
</div>
<div id="syntenic-block-rearrangement-analysis" class="section level2">
<h2>Syntenic Block Rearrangement Analysis</h2>
<p>This code determines the order of the 73 syntenic subregions in the different arrangements.<br />
Input Files in the data folder<br />
1. AR_Syntenic_Regions.txt - region order in the reference genome<br />
2. ST_Ord.txt - subregion order in ST relative to AR<br />
3. PP_Ord.txt - subregion order in PP relative to ST<br />
4. HY_Ord.txt - subregion order in HY relative to ST<br />
5. SC_Ord.txt - subregion order in SC relative to HY<br />
6. CU_Ord.txt - subregion order in CU relative to SC<br />
7. CH_Ord.txt - subregion order in CH relative to SC<br />
8. TL_Ord.txt - subregion order in TL relative to SC</p>
<p>Output Files in output folder<br />
1. AR_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate<br />
2. ST_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate<br />
3. PP_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate<br />
1. HY_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate<br />
1. SC_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate<br />
1. CU_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate<br />
1. CH_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate<br />
1. TL_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate</p>
<pre class="r"><code>#AR is a table with columns V1-V5, where
#V1 - subregion = AR_SReg
#V2 - beginning cytogenetic region
#V3 - end cytogenetic region
#V4 - beginning nucleotide coordinate in the Arrowhead (AR) arrangement = AR_Beg
#v5 - end nucleotide coordinate in the Arrowhead (AR) arrangement = AR_End
AR &lt;- read.delim(&quot;data/AR_Syntenic_Regions.txt&quot;, header = FALSE, sep = &quot;\t&quot;)
AR_SReg &lt;-AR$V1
AR_Beg &lt;- AR$V4
AR_End &lt;- AR$V5
AR_Data &lt;- cbind(AR_SReg, AR_Beg, AR_End)
write.csv(AR_Data, file =&quot;output/AR_Coordinates.csv&quot;)
#AR_lenreg - length of the subregion in the Arrowhead (AR) arrangement
AR_lenreg &lt;- AR_End - AR_Beg +1
#nsreg - number of subregions
nsreg &lt;- nrow(AR)
#
#Standard Arrangement
#
#Obtain the coordinates of the subregions in the Standard (ST) arrangement
#ST_Ord -  order of the subregions in the Standard (ST) arrangement
#ST_Beg -  beginning nucleotide coordinate in the Standard (ST) arrangement
#ST_End -  end nucleotide coordinate in the Standard (ST) arrangement
#ST_SReg - subregion in the Standard (ST) arrangement
ST_Ord &lt;- scan(&quot;data/ST_Ord.txt&quot;)
ST_Beg &lt;- c(1)
ST_End &lt;- c(AR_End[1])
ST_SReg &lt;- c(AR_SReg[1])
for (i in 2:nsreg) {
  ST_Beg[i] &lt;- ST_End[i-1]+1
  ST_End[i] &lt;- ST_End[i-1]+AR_lenreg[ST_Ord[i]]
  ST_SReg[i] &lt;- c(AR_SReg[ST_Ord[i]])
}
ST_lenreg &lt;- ST_End - ST_Beg +1
ST_Data &lt;- cbind(ST_SReg, ST_Beg, ST_End)
write.csv(ST_Data, file =&quot;output/ST_Coordinates.csv&quot;)
#
#Pikes Peak Arrangement
#
#Obtain the coordinates of the subregions in the Pikes Peak (PP) arrangement
#PP_Ord -  order of the subregions in the Pikes Peak (PP) arrangement
#PP_Beg -  beginning nucleotide coordinate in the Pikes Peak (PP) arrangement
#PP_End -  end nucleotide coordinate in the Pikes Peak (PP) arrangement
#PP_SReg - subregion in the Pikes Peak (P) arrangement
PP_Ord &lt;- scan(&quot;data/PP_Ord.txt&quot;)
PP_Beg &lt;- c(1)
PP_End &lt;- c(ST_End[1])
PP_SReg &lt;- c(ST_SReg[1])
for (i in 2:nsreg) {
  PP_Beg[i] &lt;- PP_End[i-1]+1
  PP_End[i] &lt;- PP_End[i-1]+ST_lenreg[PP_Ord[i]]
  PP_SReg[i] &lt;- c(ST_SReg[PP_Ord[i]])
}
PP_Data &lt;- cbind(PP_SReg, PP_Beg, PP_End)
write.csv(PP_Data, file =&quot;output/PP_Coordinates.csv&quot;)
#
#Hypothetical Arrangement - Ancestral Arrangement
#
#Obtain the coordinates of the subregions in the Hypothetical (HY) arrangement
#HY_Ord -  order of the subregions in the Hypothetical (HY) arrangement
#HY_Beg -  beginning nucleotide coordinate in the Hypothetical (HY) arrangement
#HY_End -  end nucleotide coordinate in the Hypothetical (HY) arrangement
#HY_SReg - subregion in the Hypothetical (HY) arrangement
HY_Ord &lt;- scan(&quot;data/HY_Ord.txt&quot;)
HY_Beg &lt;- c(1)
HY_End &lt;- c(ST_End[1])
HY_SReg &lt;- c(ST_SReg[1])
for (i in 2:nsreg) {
  HY_Beg[i] &lt;- HY_End[i-1]+1
  HY_End[i] &lt;- HY_End[i-1]+ST_lenreg[HY_Ord[i]]
  HY_SReg[i] &lt;- c(ST_SReg[HY_Ord[i]])
}
HY_lenreg &lt;- HY_End - HY_Beg +1
HY_Data &lt;- cbind(HY_SReg, HY_Beg, HY_End)
write.csv(HY_Data, file =&quot;output/HY_Coordinates.csv&quot;)
#
#Santa Cruz Arrangement
#
#Obtain the coordinates of the subregions in the Santa Cruz (SC) arrangement
#SC_Ord -  order of the subregions in the Santa Cruz (SC) arrangement
#SC_Beg -  beginning nucleotide coordinate in the Santa Cruz (SC) arrangement
#SC_End -  end nucleotide coordinate in the Santa Cruz (SC) arrangement
#SC_SReg - subregion in the Santa Cruz (SC) arrangement
SC_Ord &lt;- scan(&quot;data/SC_Ord.txt&quot;)
SC_Beg &lt;- c(1)
SC_End &lt;- c(HY_End[1])
SC_SReg &lt;- c(HY_SReg[1])
for (i in 2:nsreg) {
  SC_Beg[i] &lt;- SC_End[i-1]+1
  SC_End[i] &lt;- SC_End[i-1]+HY_lenreg[SC_Ord[i]]
  SC_SReg[i] &lt;- c(HY_SReg[SC_Ord[i]])
}
SC_lenreg &lt;- SC_End - SC_Beg +1
SC_Data &lt;- cbind(SC_SReg, SC_Beg, SC_End)
write.csv(SC_Data, file =&quot;output/SC_Coordinates.csv&quot;)
#
#Cuernavaca Arrangement
#
#Obtain the coordinates of the subregions in the Cuernavaca (CU) arrangement
#CU_Ord -  order of the subregions in the Cuernavaca (CU) arrangement
#CU_Beg -  beginning nucleotide coordinate in the Cuernavaca (CU) arrangement
#CU_End -  end nucleotide coordinate in the Cuernavaca (CU) arrangement
#CU_SReg - subregion in the Cuernavaca (CU) arrangement
CU_Ord &lt;- scan(&quot;data/CU_Ord.txt&quot;)
CU_Beg &lt;- c(1)
CU_End &lt;- c(SC_End[1])
CU_SReg &lt;- c(SC_SReg[1])
for (i in 2:nsreg) {
  CU_Beg[i] &lt;- CU_End[i-1]+1
  CU_End[i] &lt;- CU_End[i-1]+SC_lenreg[CU_Ord[i]]
  CU_SReg[i] &lt;- c(SC_SReg[CU_Ord[i]])
}
CU_Data &lt;- cbind(CU_SReg, CU_Beg, CU_End)
write.csv(CU_Data, file =&quot;output/CU_Coordinates.csv&quot;)
#
#Chiricahua Arrangement
#
#Obtain the coordinates of the subregions in the Chiricahua (CH) arrangement
#CH_Ord -  order of the subregions in the Chiricahua (CH) arrangement
#CH_Beg -  beginning nucleotide coordinate in the Chiricahua (CH) arrangement
#CH_End -  end nucleotide coordinate in the Chiricahua (CH) arrangement
#CH_SReg - subregion in the Chiricahua (CH) arrangement
CH_Ord &lt;- scan(&quot;data/CH_Ord.txt&quot;)
CH_Beg &lt;- c(1)
CH_End &lt;- c(SC_End[1])
CH_SReg &lt;- c(SC_SReg[1])
for (i in 2:nsreg) {
  CH_Beg[i] &lt;- CH_End[i-1]+1
  CH_End[i] &lt;- CH_End[i-1]+SC_lenreg[CH_Ord[i]]
  CH_SReg[i] &lt;- c(SC_SReg[CH_Ord[i]])
}
CH_Data &lt;- cbind(CH_SReg, CH_Beg, CH_End)
write.csv(CH_Data, file =&quot;output/CH_Coordinates.csv&quot;)
#
#Tree Line Arrangement
#
#Obtain the coordinates of the subregions in the Tree Line (TL) arrangement
#TL_Ord -  order of the subregions in the Tree Line (TL) arrangement
#TL_Beg -  beginning nucleotide coordinate in the Tree Line (TL) arrangement
#TL_End -  end nucleotide coordinate in the Tree Line (TL) arrangement
#TL_SReg - subregion in the Tree Line (TL) arrangement
TL_Ord &lt;- scan(&quot;data/TL_Ord.txt&quot;)
TL_Beg &lt;- c(1)
TL_End &lt;- c(SC_End[1])
TL_SReg &lt;- c(SC_SReg[1])
for (i in 2:nsreg) {
  TL_Beg[i] &lt;- TL_End[i-1]+1
  TL_End[i] &lt;- TL_End[i-1]+SC_lenreg[TL_Ord[i]]
  TL_SReg[i] &lt;- c(SC_SReg[TL_Ord[i]])
}
TL_Data &lt;- cbind(TL_SReg, TL_Beg, TL_End)
write.csv(TL_Data, file =&quot;output/TL_Coordinates.csv&quot;)</code></pre>
</div>
<div id="concatenate-dnasp-gene-conversion-data" class="section level2">
<h2>Concatenate DnaSP Gene Conversion Data</h2>
<p>This PERL script takes data from DnaSP output from the Gene Conversion analysis and creates:<br />
1. GeneConversion_Tracts.tsv - a table separated value file with conversion tracts<br />
2. GeneConversion_Stats.tsv - a table separated value file with summary statistics on conversion tracts<br />
3. GeneConversion_Sites_Regxx.tsv for the xx regions [1-14]<br />
4. GeneConversion_Check.tsv - a table separated value file that checks the input file name matches the contents of the file</p>
<pre class="perl"><code>#Extract DNASP Gene Conversion
#
#Extracts the output from the Analysis: Gene Conversion implementation of the Betran et al. (1997) analysis
#
#Written by Stephen W. Schaeffer
#Date: 7 November 2018
#
my($file,$filein,$fileout,$fileinp,$filec,$inputID,$regi,$analID);
#
#$file    - Input file name from the file of file names (fof)
#$filein  - File name of the text file with the DNASP output
#$fileout - 
#$fileinp - File name of the individual DnaSP Gene Conversion pairwise comparison
#$filec   - File name from line 8 within the DnaSP Gene Conversion file that was used to generate
#           $fileinp, which has the subregion [01a-14h]
#$inputID - &quot;Reg_Arr1_Arr2&quot; from the input file name
#$analID  - &quot;Reg_Arr1_Arr2&quot; from within the analysis file
#
#
my ($i,$line,@GCtsites,$GCtsites,@GCpop1,$GCpop1,@GCp1sam,$GCp1sam,@GCpop2,$GCpop2,@GCp2sam,$GCp2sam);
my (@GCNo,$GCNo,@GCint,$GCint,@GCstr,$GCstr,@GCcoor,$GCcoor,@GClen,$GClen);
my (@GCpsi,$GCpsi,$GCtlen,@GCsites,$GCsites,@GCSiteI,$GCSiteI,$rego,$reg);
chdir(&quot;data&quot;);
open (FILE_LIST, &quot;file_list.fof&quot;);
#
#OUTPUT1 - GeneConversion_Tracts.tsv
#is a tab separated values file that lists all gene conversion tracts from
#all pairwise comparisons of the six gene arrangements. Columns [GC_Event, number of
#the event; GC_File, filename with the data; GC_Strain, strain with the gene conversion
#event; GC_Beg, first nucleotide of the gene conversion tract; GC_End, last
#nucleotide of the gene conversion tract; GC_Len, length of the gene conversion tract]. 
#
open(OUTPUT1, &quot;&gt;GeneConversion_Tracts.tsv&quot;);
print OUTPUT1 &quot;GC_Event\tGC_File\tGC_Strain\tGC_Beg\tGC_End\tGC_Len\n&quot;;
#
#OUTPUT2 - GeneConversion_Stats.tsv
#is a tab separated values file that lists the statistics for the 15 pairwise
#comparisons of the six gene arrangements for the 73 regions for a total of 1,095
#analyses.  Columns [GC_File, filename with the data; GC_Pop1, gene arrangement 1;
#GC_Pop1_No, number of gene arrangement 1; GC_Pop2, gene arrangement 2; GC_Pop2_No,
#number of gene arrangement 2; GC_Tract_No, number of gene conversion tracts; GC_Sites,
#number of informative sites; GC_Psi, mean value of Psi for the informative sites;
#GC_Phi, mean value of Phi for the observed data].
#
open(OUTPUT2, &quot;&gt;GeneConversion_Stats.tsv&quot;);
print OUTPUT2 &quot;GC_File\tGC_Pop1\tGC_Pop1_No\tGC_Pop2\tGC_Pop2_No\tGC_Tract_No\tGC_Sites\tGC_Psi\tGC_Phi\n&quot;;
#
#OUTPUT3 - GeneConversion_Sites_Regxx.tsv
#is a tab separated values file that lists the informative sites used in the
#gene conversion detection analysis. Columns [GC_File, filename with the data; GC_Pop1,
#gene arrangement 1; GC_Pop2, gene arrangement 2; GC_SiteNo, nucleotide position of the
#informative site in the data subset; GC_SitePsi, value of Psi for the informative site;
#GC_TotalSites, total number of nucleotide sites].
#
#
#Output4 - GeneConversion_Check.tsv
#is a tab separated values file that lists the name of the file, the name
#of the Input Data File on line 8, the first arrangement and the second arrangement
#[Input_File,Analysis_File,Check], where Input_File and Analysis_File are of the form &quot;Subregion_ARR1_ARR2&quot;
#This file checks to make sure the name of the file matches with the contents of
#the file.
#
open(OUTPUT4, &quot;&gt;GeneConversion_Check.tsv&quot;);
print OUTPUT4 &quot;Input_File\tAnalysis_File\tCheck\n&quot;;
$rego=&quot;00&quot;;
while ($file = &lt;FILE_LIST&gt;) {
    chop($file);
    $fileinp=$file;
    $inputID=substr $fileinp, 15, 9;
    $reg=substr $file, 15, 2;
    if($reg ne $rego) {
        $fileout = join &quot;&quot;,&quot;GeneConversion_Sites_Reg&quot;,$reg, &quot;.tsv&quot;;
        open(OUTPUT3, &quot;&gt;&gt;&quot; ,&quot;$fileout&quot;);
        print OUTPUT3 &quot;GC_File\tGC_Pop1\tGC_Pop2\tGC_SiteNo\tGC_SitePsi\tTotalSites\n&quot;;
        close(OUTPUT3);
        $rego=$reg
    }
     $filein = join &quot;&quot;,$file,&quot;.txt&quot;;
    open (QUERY, $filein);
    for($i=0; $i&lt;=7; $i++) {
        $line=&lt;QUERY&gt;;    
    }
    #Line 8 has the filename Chr3_Syn_Reg_xxx.nex where xxx is one subregion from 01a to 14h
    chop($line);
    $filec=$line;
    #regi - the name of the region on the input file [01a-14h]
    $regi=substr $filec, 38, 3;
    #Line 9 no data used
    $line=&lt;QUERY&gt;;
    #Line 10 has the Total Number of Sites
    $line=&lt;QUERY&gt;;
    chop($line);
    #$GCtsites[1] - Total Number of Sites 
    @GCtsites = split(&#39;:&#39;, $line);
    #Line 11 no data used
    $line=&lt;QUERY&gt;;    
    #Line 12 has Population 1 Name
    $line=&lt;QUERY&gt;;    
    chop($line);
    #$GCpop1[1] - Population 1 Name 
    @GCpop1 = split(&#39;: &#39;, $line);
    #Line 13 has Population 1 Sample Size
    $line=&lt;QUERY&gt;;
    chop($line);
    #$GCp1sam[1] - Population 1 Sample Size 
    @GCp1sam = split(&#39;: &#39;, $line);
    #Line 14 has Population 2 Name
    $line=&lt;QUERY&gt;;
    chop($line);
    #$GCpop2[1] - Population 2 Name
    @GCpop2 = split(&#39;: &#39;, $line);
    #Line 15 has Population 2 Sample Size
    $line=&lt;QUERY&gt;;
    chop($line);
    #$GCp2sam[1] - Population 2 Sample Size  
    @GCp2sam = split(&#39;: &#39;, $line);
    $analID=join &quot;&quot;,$regi,&quot;_&quot;,$GCpop1[1],&quot;_&quot;,$GCpop2[1];
    if($inputID eq $analID){
      print OUTPUT4 &quot;$inputID\t$analID\tMatch\n&quot;;
    }
    else {
      print OUTPUT4 &quot;$inputID\t$analID\tNo Match\n&quot;;
    }
    #Line 16 no data used
    $line=&lt;QUERY&gt;;
    #Line 17 has the Number of Gene Conversion Tracts
    $line=&lt;QUERY&gt;;
    chop($line);
    #$GCNo[1] - Number of Gene Conversion Tracts 
    @GCNo = split(&#39;:&#39;, $line);
    #Read the conversion version tract information for the $GCNo[] Gene Conversion Tracts
    for($i=1; $i&lt;=$GCNo[1]; $i++) {
        $line=&lt;QUERY&gt;;
        chop($line);
        @GCint = split(&#39;:&#39;, $line);
        #$GCstr[0] - Strain name
        @GCstr = split(&#39; &#39;, $GCint[1]);
        #$GCcoor[0] - Beginning coordinate
        #$GCcoor[1] - End coordinate
        @GCcoor = split(&#39;-&#39;, $GCint[2]);
        $line=&lt;QUERY&gt;;
        chop($line);
        #$GClen[1] - Gene conversion tract length
        @GClen = split(&#39;:&#39;, $line);
        print OUTPUT1 &quot;$i\t$file\t$GCstr[0]\t$GCcoor[0]\t$GCcoor[1]\t$GClen[1]\n&quot;;
        }
    $line=&lt;QUERY&gt;;
    #Line 17 + (2 x $GCNo[]) + 1 has the estimate of Psi
    $line=&lt;QUERY&gt;;
    chop($line);
    #$GCpsi[1] - Psi or probability of the site being informative
    @GCpsi = split(&#39;:&#39;, $line);
    #$GCtlen - Gene conversion tract length
    $GCtlen =1/$GCpsi[1];
    #Line 17 + (2 x $GCNo[]) + 2 has the number of informative sites
    $line=&lt;QUERY&gt;;
    chop($line);
    #$GCsites[1] - Number of informative sites
    @GCsites = split(&#39;:&#39;, $line);
    print OUTPUT2 &quot;$file\t&quot;;
    print OUTPUT2 &quot;$GCpop1[1]\t$GCp1sam[1]\t&quot;;
    print OUTPUT2 &quot;$GCpop2[1]\t$GCp2sam[1]\t&quot;;
    print OUTPUT2 &quot;$GCNo[1]\t$GCsites[1]\t$GCpsi[1]\t$GCtlen\n&quot;;
    for($i=1; $i&lt;=5; $i++) {
        $line=&lt;QUERY&gt;     
        }
    $reg=substr $file, 15, 2;
    $fileout = join &quot;&quot;,&quot;GeneConversion_Sites_Reg&quot;,$reg, &quot;.tsv&quot;;
    open(OUTPUT3, &quot;&gt;&gt;&quot; ,&quot;$fileout&quot;);
    for($i=1; $i&lt;=$GCsites[1]; $i++) {
        $line=&lt;QUERY&gt;;
        chop($line);
        #$GCSiteI[0] - Site number
        #$GCSiteI[1] - Site Psi
        @GCSiteI = split(&#39; &#39;, $line);
            print OUTPUT3 &quot;$file\t&quot;;
            print OUTPUT3 &quot;$GCpop1[1]\t$GCpop2[1]\t&quot;;
            print OUTPUT3 &quot;$GCSiteI[0]\t$GCSiteI[1]\t&quot;;
            print OUTPUT3 &quot;$GCtsites[1]\n&quot;;
    }
    close(OUTPUT3);
}</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5       rstudioapi_0.11  whisker_0.4      knitr_1.30      
 [5] magrittr_1.5     R6_2.4.1         rlang_0.4.7      stringr_1.4.0   
 [9] tools_4.0.2      xfun_0.17        git2r_0.27.1     htmltools_0.5.0 
[13] ellipsis_0.3.1   rprojroot_1.3-2  yaml_2.2.1       digest_0.6.25   
[17] tibble_3.0.3     lifecycle_0.2.0  crayon_1.3.4     later_1.1.0.1   
[21] vctrs_0.3.4      promises_1.1.1   fs_1.5.0         glue_1.4.2      
[25] evaluate_0.14    rmarkdown_2.3    stringi_1.5.3    compiler_4.0.2  
[29] pillar_1.4.6     backports_1.1.10 httpuv_1.5.4     pkgconfig_2.0.3 </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
