---
title: "GeneConversionInversion"
author: "SWSchaeffer"
date: "2020-10-03"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Overview
## Syntenic Block Rearrangement Analysis
This code determines the order of the 73 syntenic subregions in the different arrangements.  
Input Files in the data folder   
1. AR_Syntenic_Regions.txt - region order in the reference genome  
2. ST_Ord.txt - subregion order in ST relative to AR  
3. PP_Ord.txt - subregion order in PP relative to ST  
4. HY_Ord.txt - subregion order in HY relative to ST  
5. SC_Ord.txt - subregion order in SC relative to HY  
6. CU_Ord.txt - subregion order in CU relative to SC  
7. CH_Ord.txt - subregion order in CH relative to SC  
8. TL_Ord.txt - subregion order in TL relative to SC  
  
Output Files in output folder  
1. AR_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate  
2. ST_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate  
3. PP_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate  
1. HY_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate  
1. SC_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate  
1. CU_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate  
1. CH_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate  
1. TL_coordinates.csv - subregion, beginning nucleotide coordinate, end nucleotide coordinate  

```{r, Syntenic Blocks}
#AR is a table with columns V1-V5, where
#V1 - subregion = AR_SReg
#V2 - beginning cytogenetic region
#V3 - end cytogenetic region
#V4 - beginning nucleotide coordinate in the Arrowhead (AR) arrangement = AR_Beg
#v5 - end nucleotide coordinate in the Arrowhead (AR) arrangement = AR_End
AR <- read.delim("data/AR_Syntenic_Regions.txt", header = FALSE, sep = "\t")
AR_SReg <-AR$V1
AR_Beg <- AR$V4
AR_End <- AR$V5
AR_Data <- cbind(AR_SReg, AR_Beg, AR_End)
write.csv(AR_Data, file ="output/AR_Coordinates.csv")
#AR_lenreg - length of the subregion in the Arrowhead (AR) arrangement
AR_lenreg <- AR_End - AR_Beg +1
#nsreg - number of subregions
nsreg <- nrow(AR)
#
#Standard Arrangement
#
#Obtain the coordinates of the subregions in the Standard (ST) arrangement
#ST_Ord -  order of the subregions in the Standard (ST) arrangement
#ST_Beg -  beginning nucleotide coordinate in the Standard (ST) arrangement
#ST_End -  end nucleotide coordinate in the Standard (ST) arrangement
#ST_SReg - subregion in the Standard (ST) arrangement
ST_Ord <- scan("data/ST_Ord.txt")
ST_Beg <- c(1)
ST_End <- c(AR_End[1])
ST_SReg <- c(AR_SReg[1])
for (i in 2:nsreg) {
  ST_Beg[i] <- ST_End[i-1]+1
  ST_End[i] <- ST_End[i-1]+AR_lenreg[ST_Ord[i]]
  ST_SReg[i] <- c(AR_SReg[ST_Ord[i]])
}
ST_lenreg <- ST_End - ST_Beg +1
ST_Data <- cbind(ST_SReg, ST_Beg, ST_End)
write.csv(ST_Data, file ="output/ST_Coordinates.csv")
#
#Pikes Peak Arrangement
#
#Obtain the coordinates of the subregions in the Pikes Peak (PP) arrangement
#PP_Ord -  order of the subregions in the Pikes Peak (PP) arrangement
#PP_Beg -  beginning nucleotide coordinate in the Pikes Peak (PP) arrangement
#PP_End -  end nucleotide coordinate in the Pikes Peak (PP) arrangement
#PP_SReg - subregion in the Pikes Peak (P) arrangement
PP_Ord <- scan("data/PP_Ord.txt")
PP_Beg <- c(1)
PP_End <- c(ST_End[1])
PP_SReg <- c(ST_SReg[1])
for (i in 2:nsreg) {
  PP_Beg[i] <- PP_End[i-1]+1
  PP_End[i] <- PP_End[i-1]+ST_lenreg[PP_Ord[i]]
  PP_SReg[i] <- c(ST_SReg[PP_Ord[i]])
}
PP_Data <- cbind(PP_SReg, PP_Beg, PP_End)
write.csv(PP_Data, file ="output/PP_Coordinates.csv")
#
#Hypothetical Arrangement - Ancestral Arrangement
#
#Obtain the coordinates of the subregions in the Hypothetical (HY) arrangement
#HY_Ord -  order of the subregions in the Hypothetical (HY) arrangement
#HY_Beg -  beginning nucleotide coordinate in the Hypothetical (HY) arrangement
#HY_End -  end nucleotide coordinate in the Hypothetical (HY) arrangement
#HY_SReg - subregion in the Hypothetical (HY) arrangement
HY_Ord <- scan("data/HY_Ord.txt")
HY_Beg <- c(1)
HY_End <- c(ST_End[1])
HY_SReg <- c(ST_SReg[1])
for (i in 2:nsreg) {
  HY_Beg[i] <- HY_End[i-1]+1
  HY_End[i] <- HY_End[i-1]+ST_lenreg[HY_Ord[i]]
  HY_SReg[i] <- c(ST_SReg[HY_Ord[i]])
}
HY_lenreg <- HY_End - HY_Beg +1
HY_Data <- cbind(HY_SReg, HY_Beg, HY_End)
write.csv(HY_Data, file ="output/HY_Coordinates.csv")
#
#Santa Cruz Arrangement
#
#Obtain the coordinates of the subregions in the Santa Cruz (SC) arrangement
#SC_Ord -  order of the subregions in the Santa Cruz (SC) arrangement
#SC_Beg -  beginning nucleotide coordinate in the Santa Cruz (SC) arrangement
#SC_End -  end nucleotide coordinate in the Santa Cruz (SC) arrangement
#SC_SReg - subregion in the Santa Cruz (SC) arrangement
SC_Ord <- scan("data/SC_Ord.txt")
SC_Beg <- c(1)
SC_End <- c(HY_End[1])
SC_SReg <- c(HY_SReg[1])
for (i in 2:nsreg) {
  SC_Beg[i] <- SC_End[i-1]+1
  SC_End[i] <- SC_End[i-1]+HY_lenreg[SC_Ord[i]]
  SC_SReg[i] <- c(HY_SReg[SC_Ord[i]])
}
SC_lenreg <- SC_End - SC_Beg +1
SC_Data <- cbind(SC_SReg, SC_Beg, SC_End)
write.csv(SC_Data, file ="output/SC_Coordinates.csv")
#
#Cuernavaca Arrangement
#
#Obtain the coordinates of the subregions in the Cuernavaca (CU) arrangement
#CU_Ord -  order of the subregions in the Cuernavaca (CU) arrangement
#CU_Beg -  beginning nucleotide coordinate in the Cuernavaca (CU) arrangement
#CU_End -  end nucleotide coordinate in the Cuernavaca (CU) arrangement
#CU_SReg - subregion in the Cuernavaca (CU) arrangement
CU_Ord <- scan("data/CU_Ord.txt")
CU_Beg <- c(1)
CU_End <- c(SC_End[1])
CU_SReg <- c(SC_SReg[1])
for (i in 2:nsreg) {
  CU_Beg[i] <- CU_End[i-1]+1
  CU_End[i] <- CU_End[i-1]+SC_lenreg[CU_Ord[i]]
  CU_SReg[i] <- c(SC_SReg[CU_Ord[i]])
}
CU_Data <- cbind(CU_SReg, CU_Beg, CU_End)
write.csv(CU_Data, file ="output/CU_Coordinates.csv")
#
#Chiricahua Arrangement
#
#Obtain the coordinates of the subregions in the Chiricahua (CH) arrangement
#CH_Ord -  order of the subregions in the Chiricahua (CH) arrangement
#CH_Beg -  beginning nucleotide coordinate in the Chiricahua (CH) arrangement
#CH_End -  end nucleotide coordinate in the Chiricahua (CH) arrangement
#CH_SReg - subregion in the Chiricahua (CH) arrangement
CH_Ord <- scan("data/CH_Ord.txt")
CH_Beg <- c(1)
CH_End <- c(SC_End[1])
CH_SReg <- c(SC_SReg[1])
for (i in 2:nsreg) {
  CH_Beg[i] <- CH_End[i-1]+1
  CH_End[i] <- CH_End[i-1]+SC_lenreg[CH_Ord[i]]
  CH_SReg[i] <- c(SC_SReg[CH_Ord[i]])
}
CH_Data <- cbind(CH_SReg, CH_Beg, CH_End)
write.csv(CH_Data, file ="output/CH_Coordinates.csv")
#
#Tree Line Arrangement
#
#Obtain the coordinates of the subregions in the Tree Line (TL) arrangement
#TL_Ord -  order of the subregions in the Tree Line (TL) arrangement
#TL_Beg -  beginning nucleotide coordinate in the Tree Line (TL) arrangement
#TL_End -  end nucleotide coordinate in the Tree Line (TL) arrangement
#TL_SReg - subregion in the Tree Line (TL) arrangement
TL_Ord <- scan("data/TL_Ord.txt")
TL_Beg <- c(1)
TL_End <- c(SC_End[1])
TL_SReg <- c(SC_SReg[1])
for (i in 2:nsreg) {
  TL_Beg[i] <- TL_End[i-1]+1
  TL_End[i] <- TL_End[i-1]+SC_lenreg[TL_Ord[i]]
  TL_SReg[i] <- c(SC_SReg[TL_Ord[i]])
}
TL_Data <- cbind(TL_SReg, TL_Beg, TL_End)
write.csv(TL_Data, file ="output/TL_Coordinates.csv")
```


## Concatenate DnaSP Gene Conversion Data
This PERL script takes data from DnaSP output from the Gene Conversion analysis and creates:  
1. GeneConversion_Tracts.tsv - a table separated value file with conversion tracts  
2. GeneConversion_Stats.tsv - a table separated value file with summary statistics on conversion tracts  
3. GeneConversion_Sites_Regxx.tsv for the xx regions [1-14]  
4. GeneConversion_Check.tsv - a table separated value file that checks the input file name matches the contents of the file  

```{perl, Concatenate DnaSP Gene Conversion Data, engine.path = 'C:\\perl64\\bin\\perl.exe'}
#Extract DNASP Gene Conversion
#
#Extracts the output from the Analysis: Gene Conversion implementation of the Betran et al. (1997) analysis
#
#Written by Stephen W. Schaeffer
#Date: 7 November 2018
#
my($file,$filein,$fileout,$fileinp,$filec,$inputID,$regi,$analID);
#
#$file    - Input file name from the file of file names (fof)
#$filein  - File name of the text file with the DNASP output
#$fileout - 
#$fileinp - File name of the individual DnaSP Gene Conversion pairwise comparison
#$filec   - File name from line 8 within the DnaSP Gene Conversion file that was used to generate
#           $fileinp, which has the subregion [01a-14h]
#$inputID - "Reg_Arr1_Arr2" from the input file name
#$analID  - "Reg_Arr1_Arr2" from within the analysis file
#
#
my ($i,$line,@GCtsites,$GCtsites,@GCpop1,$GCpop1,@GCp1sam,$GCp1sam,@GCpop2,$GCpop2,@GCp2sam,$GCp2sam);
my (@GCNo,$GCNo,@GCint,$GCint,@GCstr,$GCstr,@GCcoor,$GCcoor,@GClen,$GClen);
my (@GCpsi,$GCpsi,$GCtlen,@GCsites,$GCsites,@GCSiteI,$GCSiteI,$rego,$reg);
chdir("data");
open (FILE_LIST, "file_list.fof");
#
#OUTPUT1 - GeneConversion_Tracts.tsv
#is a tab separated values file that lists all gene conversion tracts from
#all pairwise comparisons of the six gene arrangements. Columns [GC_Event, number of
#the event; GC_File, filename with the data; GC_Strain, strain with the gene conversion
#event; GC_Beg, first nucleotide of the gene conversion tract; GC_End, last
#nucleotide of the gene conversion tract; GC_Len, length of the gene conversion tract]. 
#
open(OUTPUT1, ">GeneConversion_Tracts.tsv");
print OUTPUT1 "GC_Event\tGC_File\tGC_Strain\tGC_Beg\tGC_End\tGC_Len\n";
#
#OUTPUT2 - GeneConversion_Stats.tsv
#is a tab separated values file that lists the statistics for the 15 pairwise
#comparisons of the six gene arrangements for the 73 regions for a total of 1,095
#analyses.  Columns [GC_File, filename with the data; GC_Pop1, gene arrangement 1;
#GC_Pop1_No, number of gene arrangement 1; GC_Pop2, gene arrangement 2; GC_Pop2_No,
#number of gene arrangement 2; GC_Tract_No, number of gene conversion tracts; GC_Sites,
#number of informative sites; GC_Psi, mean value of Psi for the informative sites;
#GC_Phi, mean value of Phi for the observed data].
#
open(OUTPUT2, ">GeneConversion_Stats.tsv");
print OUTPUT2 "GC_File\tGC_Pop1\tGC_Pop1_No\tGC_Pop2\tGC_Pop2_No\tGC_Tract_No\tGC_Sites\tGC_Psi\tGC_Phi\n";
#
#OUTPUT3 - GeneConversion_Sites_Regxx.tsv
#is a tab separated values file that lists the informative sites used in the
#gene conversion detection analysis. Columns [GC_File, filename with the data; GC_Pop1,
#gene arrangement 1; GC_Pop2, gene arrangement 2; GC_SiteNo, nucleotide position of the
#informative site in the data subset; GC_SitePsi, value of Psi for the informative site;
#GC_TotalSites, total number of nucleotide sites].
#
#
#Output4 - GeneConversion_Check.tsv
#is a tab separated values file that lists the name of the file, the name
#of the Input Data File on line 8, the first arrangement and the second arrangement
#[Input_File,Analysis_File,Check], where Input_File and Analysis_File are of the form "Subregion_ARR1_ARR2"
#This file checks to make sure the name of the file matches with the contents of
#the file.
#
open(OUTPUT4, ">GeneConversion_Check.tsv");
print OUTPUT4 "Input_File\tAnalysis_File\tCheck\n";
$rego="00";
while ($file = <FILE_LIST>) {
    chop($file);
    $fileinp=$file;
    $inputID=substr $fileinp, 15, 9;
    $reg=substr $file, 15, 2;
    if($reg ne $rego) {
        $fileout = join "","GeneConversion_Sites_Reg",$reg, ".tsv";
        open(OUTPUT3, ">>" ,"$fileout");
        print OUTPUT3 "GC_File\tGC_Pop1\tGC_Pop2\tGC_SiteNo\tGC_SitePsi\tTotalSites\n";
        close(OUTPUT3);
        $rego=$reg
    }
     $filein = join "",$file,".txt";
    open (QUERY, $filein);
    for($i=0; $i<=7; $i++) {
        $line=<QUERY>;    
    }
    #Line 8 has the filename Chr3_Syn_Reg_xxx.nex where xxx is one subregion from 01a to 14h
    chop($line);
    $filec=$line;
    #regi - the name of the region on the input file [01a-14h]
    $regi=substr $filec, 38, 3;
    #Line 9 no data used
    $line=<QUERY>;
    #Line 10 has the Total Number of Sites
    $line=<QUERY>;
    chop($line);
    #$GCtsites[1] - Total Number of Sites 
    @GCtsites = split(':', $line);
    #Line 11 no data used
    $line=<QUERY>;    
    #Line 12 has Population 1 Name
    $line=<QUERY>;    
    chop($line);
    #$GCpop1[1] - Population 1 Name 
    @GCpop1 = split(': ', $line);
    #Line 13 has Population 1 Sample Size
    $line=<QUERY>;
    chop($line);
    #$GCp1sam[1] - Population 1 Sample Size 
    @GCp1sam = split(': ', $line);
    #Line 14 has Population 2 Name
    $line=<QUERY>;
    chop($line);
    #$GCpop2[1] - Population 2 Name
    @GCpop2 = split(': ', $line);
    #Line 15 has Population 2 Sample Size
    $line=<QUERY>;
    chop($line);
    #$GCp2sam[1] - Population 2 Sample Size  
    @GCp2sam = split(': ', $line);
    $analID=join "",$regi,"_",$GCpop1[1],"_",$GCpop2[1];
    if($inputID eq $analID){
      print OUTPUT4 "$inputID\t$analID\tMatch\n";
    }
    else {
      print OUTPUT4 "$inputID\t$analID\tNo Match\n";
    }
    #Line 16 no data used
    $line=<QUERY>;
    #Line 17 has the Number of Gene Conversion Tracts
    $line=<QUERY>;
    chop($line);
    #$GCNo[1] - Number of Gene Conversion Tracts 
    @GCNo = split(':', $line);
    #Read the conversion version tract information for the $GCNo[] Gene Conversion Tracts
    for($i=1; $i<=$GCNo[1]; $i++) {
        $line=<QUERY>;
        chop($line);
        @GCint = split(':', $line);
        #$GCstr[0] - Strain name
        @GCstr = split(' ', $GCint[1]);
        #$GCcoor[0] - Beginning coordinate
        #$GCcoor[1] - End coordinate
        @GCcoor = split('-', $GCint[2]);
        $line=<QUERY>;
        chop($line);
        #$GClen[1] - Gene conversion tract length
        @GClen = split(':', $line);
        print OUTPUT1 "$i\t$file\t$GCstr[0]\t$GCcoor[0]\t$GCcoor[1]\t$GClen[1]\n";
        }
    $line=<QUERY>;
    #Line 17 + (2 x $GCNo[]) + 1 has the estimate of Psi
    $line=<QUERY>;
    chop($line);
    #$GCpsi[1] - Psi or probability of the site being informative
    @GCpsi = split(':', $line);
    #$GCtlen - Gene conversion tract length
    $GCtlen =1/$GCpsi[1];
    #Line 17 + (2 x $GCNo[]) + 2 has the number of informative sites
    $line=<QUERY>;
    chop($line);
    #$GCsites[1] - Number of informative sites
    @GCsites = split(':', $line);
    print OUTPUT2 "$file\t";
    print OUTPUT2 "$GCpop1[1]\t$GCp1sam[1]\t";
    print OUTPUT2 "$GCpop2[1]\t$GCp2sam[1]\t";
    print OUTPUT2 "$GCNo[1]\t$GCsites[1]\t$GCpsi[1]\t$GCtlen\n";
    for($i=1; $i<=5; $i++) {
        $line=<QUERY>     
        }
    $reg=substr $file, 15, 2;
    $fileout = join "","GeneConversion_Sites_Reg",$reg, ".tsv";
    open(OUTPUT3, ">>" ,"$fileout");
    for($i=1; $i<=$GCsites[1]; $i++) {
        $line=<QUERY>;
        chop($line);
        #$GCSiteI[0] - Site number
        #$GCSiteI[1] - Site Psi
        @GCSiteI = split(' ', $line);
            print OUTPUT3 "$file\t";
            print OUTPUT3 "$GCpop1[1]\t$GCpop2[1]\t";
            print OUTPUT3 "$GCSiteI[0]\t$GCSiteI[1]\t";
            print OUTPUT3 "$GCtsites[1]\n";
    }
    close(OUTPUT3);
}
```

